gui/dig.lua

```diff
@@ -0,0 +1,658 @@
+
+-- Stretch Goals
+-----------------------------
+-- Shape preview in panel
+-- Shape designer in preview panel to draw repeatable shapes i'e' 2x3 room with door
+
+-- END TODOS ================
+
+local gui = require("gui")
+local guidm = require("gui.dwarfmode")
+local widgets = require("gui.widgets")
+local quickfort = reqscript("quickfort")
+local shapes = reqscript("internal/dig/shapes")
+
+local function get_dims(pos1, pos2)
+    local width, height, depth =
+    math.abs(pos1.x - pos2.x) + 1,
+        math.abs(pos1.y - pos2.y) + 1,
+        math.abs(pos1.z - pos2.z) + 1
+    return width, height, depth
+end
+
+-- Panel to show the Mouse position/dimensions/etc
+-- Stolen from blueprint or quickfort I forget
+ActionPanel = defclass(ActionPanel, widgets.ResizingPanel)
+ActionPanel.ATTRS {
+    get_mark_fn = DEFAULT_NIL,
+    is_setting_start_pos_fn = DEFAULT_NIL,
+    autoarrange_subviews = true,
+}
+
+function ActionPanel:init()
+    self:addviews { widgets.WrappedLabel {
+        view_id = "action_label",
+        text_to_wrap = self:callback("get_action_text"),
+    },
+        widgets.TooltipLabel {
+            view_id = "selected_area",
+            indent = 1,
+            text = { { text = self:callback("get_area_text") } },
+            show_tooltip = self.get_mark_fn,
+        } }
+end
+
+function ActionPanel:get_action_text()
+    local text = "Select the "
+    if self.get_mark_fn() then
+        text = text .. "second corner"
+    else
+        text = text .. "first corner"
+    end
+    return text .. " with the mouse."
+end
+
+function ActionPanel:get_area_text()
+    local mark = self.get_mark_fn()
+    if not mark then
+        return ""
+    end
+    local other = dfhack.gui.getMousePos() or {
+        x = mark.x,
+        y = mark.y,
+        z = df.global.window_z,
+    }
+    local width, height, depth = get_dims(mark, other)
+    local tiles = width * height * depth
+    local plural = tiles > 1 and "s" or ""
+    return ("%dx%dx%d (%d tile%s) mark: %d, %d, %d"):format(
+        width,
+        height,
+        depth,
+        tiles,
+        plural,
+        other.x,
+        other.y,
+        other.z
+    )
+end
+
+-- Generic options not specific to shapes
+GenericOptionsPanel = defclass(GenericOptionsPanel, widgets.ResizingPanel)
+GenericOptionsPanel.ATTRS {
+    name = DEFAULT_NIL,
+    autoarrange_subviews = true,
+    dig_panel = DEFAULT_NIL,
+    on_layout_change = DEFAULT_NIL,
+}
+
+function GenericOptionsPanel:init()
+    local options = {}
+    for i, shape in pairs(shapes.all_shapes) do
+        options[#options + 1] = {
+            label = shape.name,
+            value = i,
+        }
+    end
+
+    self:addviews { widgets.WrappedLabel {
+        view_id = "settings_label",
+        text_to_wrap = "General Settings:\n",
+    },
+        widgets.CycleHotkeyLabel {
+            view_id = "shape_name",
+            key = "CUSTOM_E",
```

Comment:
Oops., changed it to `z`. I've mostly stuck with the new default hotkeys to try and get used to them except I remapped up/down to <,> because I just couldn't adjust

Reply:
ack. I did that as well, actually.